gson/src/main/java/com/google/gson/JsonPrimitive.java

```diff
@@ -290,11 +290,10 @@ public boolean equals(Object obj) {
           : this.getAsNumber().longValue() == other.getAsNumber().longValue();
     }
     if (value instanceof Number && other.value instanceof Number) {
-      double a = getAsNumber().doubleValue();
-      // Java standard types other than double return true for two NaN. So, need
-      // special handling for double.
-      double b = other.getAsNumber().doubleValue();
-      return a == b || (Double.isNaN(a) && Double.isNaN(b));
+      return this.value instanceof BigDecimal && other.value instanceof BigDecimal
+          ? this.getAsBigDecimal().compareTo(other.getAsBigDecimal()) == 0
```

Comment:
Really minor, but might be good to add a comment explaining why `compareTo` was chosen instead of `equals`.
I assume it is to ignore scale so that `new BigDecimal("0")` and `new BigDecimal("0.0")` are considered equal.

```suggestion
          // Uses compareTo to ignore scale of values, e.g. `0` and `0.00` should be considered equal
          ? this.getAsBigDecimal().compareTo(other.getAsBigDecimal()) == 0
```